subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2150 @@
+	uint8_t bDescriptorSubtype;
+	uint8_t bEndpointAddress;
+	uint8_t bNumImageSizePatterns;
+	struct {
+		uint16_t wWidth;
+		uint16_t wHeight;
+	} n[1] __packed;
+	uint8_t bNumCompressionPattern;
+	uint8_t bCompression[1];
+} __packed;
+
+struct uvc_format_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bFormatIndex;
+	uint8_t bNumFrameDescriptors;
+} __packed;
+
+struct uvc_format_uncomp_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bFormatIndex;
+	uint8_t bNumFrameDescriptors;
+	uint8_t guidFormat[16];
+	uint8_t bBitsPerPixel;
+	uint8_t bDefaultFrameIndex;
+	uint8_t bAspectRatioX;
+	uint8_t bAspectRatioY;
+	uint8_t bmInterlaceFlags;
+	uint8_t bCopyProtect;
+} __packed;
+
+struct uvc_format_mjpeg_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bFormatIndex;
+	uint8_t bNumFrameDescriptors;
+	uint8_t bmFlags;
+#define UVC_MJPEG_FLAGS_FIXEDSIZESAMPLES (1 << 0)
+	uint8_t bDefaultFrameIndex;
+	uint8_t bAspectRatioX;
+	uint8_t bAspectRatioY;
+	uint8_t bmInterlaceFlags;
+	uint8_t bCopyProtect;
+} __packed;
+
+struct uvc_frame_continuous_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bFrameIndex;
+	uint8_t bmCapabilities;
+	uint16_t wWidth;
+	uint16_t wHeight;
+	uint32_t dwMinBitRate;
+	uint32_t dwMaxBitRate;
+	uint32_t dwMaxVideoFrameBufferSize;
+	uint32_t dwDefaultFrameInterval;
+	uint8_t bFrameIntervalType;
+	uint32_t dwMinFrameInterval;
+	uint32_t dwMaxFrameInterval;
+	uint32_t dwFrameIntervalStep;
+} __packed;
+
+struct uvc_frame_discrete_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bFrameIndex;
+	uint8_t bmCapabilities;
+	uint16_t wWidth;
+	uint16_t wHeight;
+	uint32_t dwMinBitRate;
+	uint32_t dwMaxBitRate;
+	uint32_t dwMaxVideoFrameBufferSize;
+	uint32_t dwDefaultFrameInterval;
+	uint8_t bFrameIntervalType;
+	uint32_t dwFrameInterval[CONFIG_USBD_VIDEO_MAX_FRMIVAL];
+} __packed;
+
+union uvc_stream_descriptor {
+	struct uvc_format_descriptor format;
+	struct uvc_format_uncomp_descriptor format_uncomp;
+	struct uvc_format_mjpeg_descriptor format_mjpeg;
+	struct uvc_frame_discrete_descriptor frame_discrete;
+	struct uvc_frame_continuous_descriptor frame_continuous;
+} __packed;
+
+struct uvc_color_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bColorPrimaries;
+	uint8_t bTransferCharacteristics;
+	uint8_t bMatrixCoefficients;
+} __packed;
+
+struct usbd_uvc_desc {
+	struct usb_association_descriptor iad;
+	struct usb_if_descriptor if0;
+	struct uvc_control_header_descriptor if0_header;
+	struct usb_desc_header nil_desc;
+};
+
+struct usbd_uvc_strm_desc {
+	struct uvc_camera_terminal_descriptor if0_ct;
+	struct uvc_selector_unit_descriptor if0_su;
+	struct uvc_processing_unit_descriptor if0_pu;
+	struct uvc_extension_unit_descriptor if0_xu;
```

Comment:
The UVC extension unit is something that could exist, but it looks like it is always here. Is there a way to have 0 or >1 extension unit?

Reply:
Reminder for anyone following this thread: previous version of this PR allowed arbitrary topology of devices, but Linux only supports one [particular topology](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/media/usb/uvc/uvc_driver.c#n1347).

It does allow an arbitrary number of extension unit (XU) though, and I can start working on it. Directly pushed to this PR if not merged when done, or a new PR.

Since there is a GUID for each XU, this needs a per-instance configuration.
The only challenge left is thinking of a devicetree or runtime API that would be intuitive.
I will try to build a small table to see what XU typically look like in advanced webcams the wild.

Reply:
Example of extension descriptors defined by a vendor found in the wild: https://github.com/Huddly/XU-API